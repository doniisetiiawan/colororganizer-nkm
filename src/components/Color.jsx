import PropTypes from 'prop-types';
import React from 'react';
import StarRating from './StarRating';

const Color = ({
  title,
  color,
  rating,
  onRemove,
  onRate,
}) => (
  <section className="color">
    <h1>{title}</h1>
    <button type="button" onClick={onRemove}>X</button>
    <div
      className="color"
      style={{ backgroundColor: color }}
    />
    <div>
      <StarRating starsSelected={rating} onRate={onRate} />
    </div>
  </section>
);

export default Color;

Color.propTypes = {
  color: PropTypes.string.isRequired,
  onRate: PropTypes.func,
  onRemove: PropTypes.func,
  rating: PropTypes.number,
  title: PropTypes.string.isRequired,
};

Color.defaultProps = {
  onRate: (f) => f,
  onRemove: (f) => f,
  rating: 0,
};
