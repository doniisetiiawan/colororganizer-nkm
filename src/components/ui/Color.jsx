import PropTypes from 'prop-types';
import React, { createRef } from 'react';
import StarRating from './StarRating';

class Color extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};
    this.title = createRef();
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount = () => {
    this.style = { backgroundColor: '#CCC' };
  };

  shouldComponentUpdate = (nextProps) => {
    const { rating } = this.props;
    return rating !== nextProps.rating;
  };

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillUpdate = (nextProps) => {
    const { title, rating } = this.props;
    this.style = null;
    alert(
      `${title}: rating ${rating} -> ${nextProps.rating}`,
    );
  };

  componentDidUpdate = (prevProps) => {
    const { title, rating } = this.props;
    const status = rating > prevProps.rating ? 'better' : 'worse';
    console.log(`${title} is getting ${status}`);
    this.title.current.style.backgroundColor = '';
    this.title.current.style.color = 'black';
  };

  render() {
    const {
      title,
      color,
      rating,
      onRemove,
      onRate,
    } = this.props;
    return (
      <section className="color" style={this.style}>
        <h1 ref={this.title}>{title}</h1>
        <button type="button" onClick={onRemove}>
          X
        </button>
        <div
          className="color"
          style={{ backgroundColor: color }}
        />
        <div>
          <StarRating
            starsSelected={rating}
            onRate={onRate}
          />
        </div>
      </section>
    );
  }
}

export default Color;

Color.propTypes = {
  color: PropTypes.string.isRequired,
  onRate: PropTypes.func,
  onRemove: PropTypes.func,
  rating: PropTypes.number,
  title: PropTypes.string.isRequired,
};

Color.defaultProps = {
  onRate: (f) => f,
  onRemove: (f) => f,
  rating: 0,
};
