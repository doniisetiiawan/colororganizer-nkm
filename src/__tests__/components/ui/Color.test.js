/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';
import { compose } from 'redux';
import Color from '../../../components/ui/Color';

configure({ adapter: new Adapter() });

describe('<Color /> UI Component', () => {
  const shallowExpect = compose(expect, toJSON, shallow);

  it('Renders correct properties', () => shallowExpect(
    <Color
      title="Test Color"
      color="#F0F0F0"
      rating={3}
      timestamp="Mon Apr 11 2016 12:54:19 GMT-0700 (PDT)"
    />,
  ).toMatchSnapshot());

  it('Invokes onRate property', () => {
    mount(<Color title="Test Color" color="#F0F0F0" />)
      .find('div.star-rating')
      .childAt(2)
      .simulate('click');
  });

  it('Invokes onRate property', () => {
    mount(<Color title="Test Color" color="#F0F0F0" />)
      .find('button')
      .simulate('click');
  });
});
