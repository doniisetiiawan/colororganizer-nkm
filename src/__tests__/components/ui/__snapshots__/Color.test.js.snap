// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Color /> UI Component Renders correct properties 1`] = `
<section
  className="color"
  style={
    Object {
      "backgroundColor": "#CCC",
    }
  }
>
  <h1>
    Test Color
  </h1>
  <button
    onClick={[Function]}
    type="button"
  >
    X
  </button>
  <div
    className="color"
    style={
      Object {
        "backgroundColor": "#F0F0F0",
      }
    }
  />
  <div>
    <StarRating
      onRate={[Function]}
      starsSelected={3}
      totalStars={5}
    />
  </div>
</section>
`;
